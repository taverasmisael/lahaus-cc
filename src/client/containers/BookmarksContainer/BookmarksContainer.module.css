.container {
  @apply grid sm:grid-cols-2 md:grid-cols-3 2xl:grid-cols-4 gap-6 md:gap-x-8 md:gap-y-14;
}
