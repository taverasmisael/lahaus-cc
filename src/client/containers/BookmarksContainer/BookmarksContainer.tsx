import BookmarkCard from '@/client/components/BookmarkCard'
import { useBookmarks } from '@/client/hooks/useBookmarks'
import { Bookmark } from '@/shared/bookmark.interface'
import styles from './BookmarksContainer.module.css'

function BookmarksContainer() {
  const state = useBookmarks(true)
  return state.isLoading ? (
    <div>Loading...</div>
  ) : (
    <ul className={styles.container}>
      {(state.bookmarks as Bookmark[]).map((bookmark) => (
        <li className={styles.item} key={bookmark.id}>
          <BookmarkCard
            items={bookmark.realEstates}
            id={bookmark.id}
            name={bookmark.name}
          />
        </li>
      ))}
    </ul>
  )
}

export default BookmarksContainer
