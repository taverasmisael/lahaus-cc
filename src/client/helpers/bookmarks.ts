import { Bookmark } from '@/shared/bookmark.interface'
import { RealEstate } from '@/shared/real-state.interface'
import { ResponseBookmarks } from '@/shared/response-bookmarks.interface'

export const normalizeResponse = (res: ResponseBookmarks): Bookmark[] => {
  const { data, included } = res

  const bookmarks = data.map((bm) => {
    const realEstates = bm.attributes.real_estate_ids.reduce((prev, id) => {
      const realEstate = included.find((state) => state.id === id.toString())
      return realEstate ? [...prev, realEstate] : prev
    }, [] as RealEstate[])

    return {
      id: bm.id,
      name: bm.attributes.name,
      realEstates,
    }
  })

  return bookmarks
}
