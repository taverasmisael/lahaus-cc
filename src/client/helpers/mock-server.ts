// Use this file ONLY on test envs

import { rest } from 'msw'
import { setupServer } from 'msw/node'

import { RealEstateAttributes } from '@/shared/real-state.interface'
import { ResponseBookmarks } from '@/shared/response-bookmarks.interface'

export const DEFAULT_RESPONSE: ResponseBookmarks = {
  data: [{ attributes: { name: 'TEST', real_estate_ids: [3] }, id: '1' }],
  included: [
    {
      id: '3',
      attributes: {
        name: 'TEST',
        gallery_urls: ['https://unsplas.it/160/190'],
      } as RealEstateAttributes,
    },
  ],
}

export const server = setupServer(
  rest.get(process.env.NEXT_PUBLIC_API_URL as string, (req, res, ctx) => {
    return res(ctx.json(DEFAULT_RESPONSE))
  })
)

export const initializeMockServer = () => {
  // Enable API mocking before tests.
  beforeAll(() => server.listen())

  // Reset any runtime request handlers we may add during the tests.
  afterEach(() => server.resetHandlers())

  // Disable API mocking after the tests are done.
  afterAll(() => server.close())
}
