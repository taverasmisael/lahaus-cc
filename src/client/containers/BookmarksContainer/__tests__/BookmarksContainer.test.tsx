import { render, waitForElementToBeRemoved } from '@testing-library/react'
import { initializeMockServer } from '@/client/helpers/mock-server'

import BookmarksContainer from '../BookmarksContainer'

initializeMockServer()

describe('containers/BookmarksContainer', () => {
  it('renders without crashing', () => {
    const { container } = render(<BookmarksContainer />)
    expect(container.firstChild).toMatchSnapshot()
  })
  it('renders the list of BookmarkCards', async () => {
    const { getByText, getByTestId, getAllByTestId, container } = render(
      <BookmarksContainer />
    )

    await waitForElementToBeRemoved(() => getByText(/loading/gi))
    expect(getByTestId('list')).toBeInTheDocument()
    expect(getAllByTestId('card')).toHaveLength(2)

    expect(container.firstChild).toMatchSnapshot()
  })
})
