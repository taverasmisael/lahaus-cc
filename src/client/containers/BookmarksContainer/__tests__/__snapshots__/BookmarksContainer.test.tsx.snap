// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`containers/BookmarksContainer renders the list of BookmarkCards 1`] = `
<div
  class="container"
  data-testid="list"
>
  <div
    class="item"
    data-testid="card"
  >
    <div
      class="container"
    >
      <header
        class="header"
      >
        <div
          class="carousel singleChild"
        >
          <div
            class="image image0"
          >
            <img
              alt="TEST"
              class="w-full h-full object-cover"
              src="https://unsplas.it/160/190"
            />
          </div>
        </div>
      </header>
      <h2
        class="title"
      >
        <a
          href="/bookmarks/1"
        >
          TEST
        </a>
      </h2>
      <p
        class="description"
      >
        1 propiedad guardada
      </p>
    </div>
  </div>
  <div
    class="item"
    data-testid="card"
  >
    <div
      class="container"
    >
      <header
        class="header"
      >
        <div
          class="carousel singleChild"
        >
          <div
            class="image image0"
          >
            <img
              class="w-full h-full object-cover"
              src="/imgs/empty-state.png"
            />
          </div>
        </div>
      </header>
      <h2
        class="title"
      >
        <a
          href="/bookmarks"
        >
          Empty list
        </a>
      </h2>
      <p
        class="description"
      >
        0 propiedades guardadas
      </p>
    </div>
  </div>
</div>
`;

exports[`containers/BookmarksContainer renders without crashing 1`] = `
<div>
  Loading...
</div>
`;
